### PR TITLE
Redirect tests

### PR DESCRIPTION
**Pull request by ScottDowne**

### CODE REVIEW

File: components/client.jsx [hunk 1]
```diff
@@ -26,14 +26,9 @@ function createElement(Component, props) {
...
 function onEnter(nextState, replaceState) {
-  var pathname = langURLParser(nextState.location);
-  var query = nextState.location.query;
-  if (pathname) {
-    if (query.redirect) {
-      pathname += query.query || "";
-      query = {};
-    }
-    replaceState({}, pathname, query);
+  var result = redirectURLParser(nextState.location);
```
Comment (ScottDowne) [inline components/client.jsx]: I moved the above functionality into langURLParser to make the functionality easily testable.

I also changed the name of the function as it's doing more than just locale redirects, but generic redirects like `en-US/nseobius` and redirects with query unrelated to the locale.

File: scripts/redirectURLParser.js [hunk 2]
```diff
@@ -1,17 +1,19 @@
...
+    redirectQuery = query.redirect;
   }
 
   // We don't need to redirect.
-  if (!redirectQuery && state.pathname !== "/") {
-    return;
+  if (!redirectQuery && location.pathname !== "/") {
+    return {
+      pathname: ""
+    };
```
Comment (ScottDowne) [inline scripts/redirectURLParser.js]: returning a more predictable value in general from this function, and trying to keep the return type the same. This is also better for testing.

File: scripts/redirectURLParser.js [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
-function langURLParser(state) {
+function redirectURLParser(location) {
   var redirectQuery = false;
-  var queryString = state.query;
+  var query = location.query;
 
-  if (queryString) {
-    redirectQuery = queryString.redirect;
+  if (query) {
+    redirectQuery = query.redirect;
```
Comment (ScottDowne) [inline scripts/redirectURLParser.js]: All the above changes are name changes.

File: scripts/redirectURLParser.js [hunk 4]
```diff
@@ -29,10 +31,16 @@ function langURLParser(state) {
 
   // invalid lang but with a valid path.
   if (!langTest && Object.keys(pages).indexOf(pathname) !== -1) {
-    return "/" + validLang + pathname;
+    pathname = "/" + validLang + pathname;
+  } else {
+    pathname = "/" + validLang + "/";
   }
-  // default
-  return "/" + validLang + "/";
```
Comment (ScottDowne) [inline scripts/redirectURLParser.js]: We are returning once at the end of the function now to reproduce less code related to query string values now being moved into this function and returned to make testing easier.

File: scripts/redirectURLParser.js [hunk 5]
```diff
@@ -29,10 +31,16 @@ function langURLParser(state) {
...
+
+  if (redirectQuery) {
+    pathname += query.query || "";
+    query = {};
+  }
+  return {pathname, query};
```
Comment (ScottDowne) [inline scripts/redirectURLParser.js]: This is the bits that used to be in client.jsx `onEnter` now moved in here as explained earlier, easier to test.

File: tests/http/redirects.test.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
+var server = require('../../server');
```
Comment (ScottDowne) [inline tests/http/redirects.test.js]: This are just some simple tests on the server for the redirects from a 404.

Comment (ScottDowne) [inline tests/http/redirects.test.js]: It's only have the picture though, as this is passed off to react router and tested separatly.

File: tests/lib/langURLParser.tests.js [hunk 7]
```diff
@@ -1,40 +0,0 @@
-var langURLParser = require('../../scripts/langURLParser.js');
```
Comment (ScottDowne) [inline tests/lib/langURLParser.tests.js]: removed via name change, not sure why github isn't figuring this one out.

File: tests/lib/redirectURLParser.tests.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
+var redirectURLParser = require('../../scripts/redirectURLParser.js');
+var should = require('should');
+
+var tests = [
```
Comment (ScottDowne) [inline tests/lib/redirectURLParser.tests.js]: I made the results much more predictable so now the test can be much more dry.

**Comment by ScottDowne** [comment]
@alicoding r? I commented in the ticket to help explain things.

**Comment by alicoding** [comment]
R+

**Pull request merged**